Improve: remember 2D mapper zoom amounts between sessions

SlySven opened the pull request:
This is intended to close #2388.

This enables the 2D (only) map zoom amounts for each map area to be save independently and restored when switching between areas in the mapper. It also saves the data between session.

It extends the existing `setMapZoom(...)` API to take an optional second argument to specify the map area ID to set the zoom (which is a floating point number) for any existing area, not just the current one.

It also adds a `getMapZoom(...)` function that, without any arguments, returns the currently used 2D map zoom value for the area currently being shown in the 2D mapper. If an area ID is provided it instead returns the value that was last used for that area - or the default value that is used initially on starting the profile or for an area that has not been viewed before.

Importantly when switching between the areas in the 2D mapper the values are retained and applied so that one area can be zoomed in and another zoomed out and switching from the first to the second and back to the first means that the zoom level used in the first is reused when it is returned to. Deleting an area will forget the stored zoom level so if it is reused it starts from scratch.

Code to save the zoom level for each area has also been implemented within the C++ core. It saves it in the Area User Data for current map formats (but removes it on loading so the user never sees it there) under a `system.fallback_map2DZoom` key but will save it directly in the binary data (which is more efficient) in the next format version whenever it is enabled.

A new Mudlet event, which has been called `mapAreaViewedChangeEvent` has been added with two additional arguments being the area ID changed to followed by the one that it was changed from. I originally thought I would need it to handle saving the zoom level for each area via the Lua system but I found that that was not practicable. I'll leave it in for now but it could be removed if peer-review finds it not to be of use.

Also, in refactoring `T2DMap::paintEvent(...)` I:
Removed/combined some locals:
* `(TArea*) playerArea` and `(TArea*) pPlayerArea` ==> `pArea`
* `(TRoom*) playerRoom` ==> `pPlayerRoom`

Remove unneeded (refactored out):
* `(qreal) ox`
* `(qreal) oy`

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Improve: remember mapper zoom amounts between sessions

This is intended to close #2388.

This enables the 2D (only) map zoom amounts for each map area to be save
independently and restored when switching between areas in the mapper. It
also saves the data between session.

It extends the existing `setMapZoom(...)` API to take an optional second
argument to specify the map area ID to set the zoom (which is a floating
point number) for any existing area, not just the current one.

It also adds a `getMapZoom(...)` function that, without any arguments,
returns the currently used 2D map zoom value for the area currently being
shown in the 2D mapper. If an area ID is provided it instead returns the
value that was last used for that area - or the default value that is used
initially on starting the profile or for an area that has not been viewed
before.

Importantly when switching between the areas in the 2D mapper the values
are retained and applied so that one area can be zoomed in and another
zoomed out and switching from the first to the second and back to the first
means that the zoom level used in the first is reused when it is returned
to. Deleting an area will forget the stored zoom level so if it is reused
it starts from scratch.

Code to save the zoom level for each area has also been implemented within
the C++ core. It saves it in the Area User Data for current map formats
(but removes it on loading so the user never sees it there) under a
`system.fallback_map2DZoom` key but will save it directly in the binary
data (which is more efficient) in the next format version whenever it is
enabled.

A new Mudlet event, which has been called `mapAreaViewedChangeEvent` has
been added with two additional arguments being the area ID changed to
followed by the one that it was changed from. I originally thought I
would need it to handle saving the zoom level for each area via the Lua
system but I found that that was not practicable. I'll leave it in for
now but it could be removed if peer-review finds it not to be of use.

Also, in refactoring `T2DMap::paintEvent(...)` I:
Removed/combined some locals:
* `(TArea*) playerArea` and `(TArea*) pPlayerArea` ==> `pArea`
* `(TRoom*) playerRoom` ==> `pPlayerRoom`

Remove unneeded (refactored out):
* `(qreal) ox`
* `(qreal) oy`

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap reviewed (approved):
This does what it says on the tin.
Ideas for further improvement:
- Also remember the previous position on map, because right now, switching back and to a map will indeed show the same zoom level, but a different position, if you moved in the intermediate area meanwhile.
- Prevent mapper zoom / pan combinations which actually show no rooms at all in the resulting view part
- Smartly calculate a "good" default zoom level for each first visit in an area, so it shows the whole map, unless it gets "too big", whatever that means exactly

Kebap reviewed (dismissed):
There seems to be something wrong, as setMapZoom(integer) does not change zoom view at all, and:

![image](https://user-images.githubusercontent.com/117238/223950680-7837f6d1-1955-42b5-867d-2088398671fe.png)

Kebap commented:
If I request `lua setMapZoom(1, nil)` this should set the current area map zoom to 1, right? 
Instead, I receive an error informing me about a missing optional argument:
`<setMapZoom: bad argument #2 type (area id as number is optional, got nil!)>`

SlySven commented:
> There seems to be something wrong, as setMapZoom(integer) does not change zoom view at all, and:
> 
> ![image](https://user-images.githubusercontent.com/117238/223950680-7837f6d1-1955-42b5-867d-2088398671fe.png)

No, the minimum zoom value is 3.0 - anything less than that will produce no effect.

vadi2 commented:
In that case it should be a Lua error since it'll never be possible to have a zoom level less than 3. What does zoom value 3.0 actually look like though?

SlySven commented:
> If I request `lua setMapZoom(1, nil)` this should set the current area map zoom to 1, right? Instead, I receive an error informing me about a missing optional argument: `<setMapZoom: bad argument #2 type (area id as number is optional, got nil!)>`

The second argument is optional, but by explicitly providing a `nil` you have provided a value - but it is not the correct type (it is a `nil` not a number) as the error message is reporting.

Remember someone did a lot of refactoring to create and use all those `getVerifiedXxxx(...)` functions in the `TLuaInterpreter` class and they included an argument in them to report that the argument that it is extracting a value for from what was passed to the Lua function that the user is invoking is "optional" rather than "required". However those same functions do not have any code in them to check that the "optional" argument ***was actually provided by the user***! :roll_eyes:

SlySven commented:
> In that case it should be a Lua error since it'll never be possible to have a zoom level less than 3. What does zoom value 3.0 actually look like though?

That isn't an (argument **type**) error, it would only be a (run-time, value) warning - and since the previous code did not do anything to report such an issue and behaved in the same way (as far as setting the zoom for the currently viewed area) I didn't feel compelled to add it.

`3.0` looks something like this on my particular setup:
![image](https://user-images.githubusercontent.com/6163092/223961037-eda73e4a-9139-4b82-a40c-56c0d2970232.png)
That is on a 1920x1080 monitor.

vadi2 commented:
For backwards compatibility that would make sense. It shouldn't return `true` though should it?

Kebap commented:
> Remember someone did a lot of refactoring  🙄

What you describe seems like a worthwhile addition (in a new PR).
Indeed I have now compared other existing functions and they show a similar flaw when you provide an optional argument with a nil value.
Then it is OK here for now as well.

SlySven commented:
> For backwards compatibility that would make sense. It shouldn't return `true` though should it?

Well, given that it has been extended - and there are a number of ways that it can fail and return a `nil` + error message I think it is not unreasonable to return a `true` on success. Given that, I now think that it would not be unreasonable to report if the zoom level supplied was too small - that would have given @Kebap a better clue as to what was going on when he tried it and got nothing back when it didn't work. Fortunately that will be a trivial change to do - especially as I parametrised(?) the minimum to a `public`ly available `static` symbol `(int) T2DMap::csmMinXYZoom`! :grinning:

SlySven committed:
Revise: add run-time (value) warning when setZoom(...) has a too small zoom

Also revise the error message for it and the corresponding getter
`getZoom(...)`, when the versions with an area id provided has an id that
doesn't exist, to the same text as some other cases in the Lua API.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -9009,16 +9009,63 @@ int TLuaInterpreter::createMapLabel(lua_State* L)
+    }
+
+    if (zoom < T2DMap::csmMinXYZoom) {
+        return warnArgumentValue(L, __func__, qsl("%1 is less than the minimum value allowed for zoom which is %2").arg(QString::number(zoom), QString::number(T2DMap::csmMinXYZoom)));
```
Align language to other warnings in this file.
```suggestion
        return warnArgumentValue(L, __func__, qsl("zoom %1 is invalid, it must be greater than %2").arg(QString::number(zoom), QString::number(T2DMap::csmMinXYZoom)));
```

SlySven replied:
Technically it should be "greater than or equal to" or "not less than" - the second is shorter in English, not sure if that negative form is the clearest to use though.

Kebap replied:
Fair enough.
```suggestion
        return warnArgumentValue(L, __func__, qsl("zoom %1 is invalid, it must be at least %2").arg(QString::number(zoom), QString::number(T2DMap::csmMinXYZoom)));
```
You're welcome.

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -9009,16 +9009,63 @@ int TLuaInterpreter::createMapLabel(lua_State* L)
+    if (!result) {
+        // Whilst it will also return false should there not be a map or mapper
+        // we should have already ruled those cases out:
+        return warnArgumentValue(L, __func__, qsl("number %1 is not a valid area id").arg(QString::number(areaID.value())));
```
as discussed in https://github.com/Mudlet/Mudlet/issues/2605
```suggestion
        return warnArgumentValue(L, __func__, qsl("number %1 is not a valid areaID").arg(QString::number(areaID.value())));
```

SlySven replied:
There are multiple instances of *both* texts in the file IIRC and I didn't do a count of each to find which one was precisely the most common... as it is that issue has not, um, been actioned/resolved yet! However I will take on board that you have pointed it out in regard to these texts in the PR.

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -9009,16 +9009,63 @@ int TLuaInterpreter::createMapLabel(lua_State* L)
+
+    if (areaID.has_value()) {
+        if (!host.mpMap->mpRoomDB->getArea(areaID.value())) {
+            return warnArgumentValue(L, __func__, qsl("number %1 is not a valid area id").arg(QString::number(areaID.value())));
```
```suggestion
            return warnArgumentValue(L, __func__, qsl("number %1 is not a valid areaID").arg(QString::number(areaID.value())));
```

Kebap commented:
The default zoom value seems not very enticing, and short of implementing any smart calculation, I implore you to reconsider this:

> > I foresee setting zoom per-area being a massive amount of pain since it'll reset between areas on the first time and so on.
> 
> If you enter a new area for the first time that does not have a zoom setting yet, start with the previous area's one.

Kebap commented:
> There seems to be something wrong, as setMapZoom(integer) does not change zoom view at all

While I now see a minimum value announced (is there also a maximum?) the zoom value still doesn't change:

![image](https://user-images.githubusercontent.com/117238/224190486-34bba749-348e-482b-ae1b-17c68fc79df7.png)

Also the area displayed in mapper does not change its zoom levels during this at all.

SlySven commented:
> The default zoom value seems not very enticing, and short of implementing any smart calculation, I implore you to reconsider this:
> 
> > > I foresee setting zoom per-area being a massive amount of pain since it'll reset between areas on the first time and so on.
> > 
> > 
> > If you enter a new area for the first time that does not have a zoom setting yet, start with the previous area's one.

Those values are the ones that are ***already*** in place, the mapper was starting off the session at 20.0 and 3.0  was the existing limit - therefore they are the least surprising values to use going forward. This PR currently means that if you go to a newly created area (or to one for the first time since you use Mudlet with this PR in place) you will get the default 20.0 zoom - because each area is now handled separately.

Whilst I suppose it might be possible to copy over the zoom from the last area previous viewed I think that that will required extra code and make this PR larger. Would it be reasonable to defer that to a separate PR?

SlySven commented:
> > There seems to be something wrong, as setMapZoom(integer) does not change zoom view at all
> 
> While I now see a minimum value announced (is there also a maximum?) the zoom value still doesn't change:
> 
> {image omitted}
> 
> Also the area displayed in mapper does not change its zoom levels during this at all.

There is no maximum limit (well other than an inherent `int` overflow!)...

Let me check that - I thought it did repaint the map in the new zoom, but maybe I left out an `update();` for the case when the area being viewed is the one that the zoom is set for; as it is now possible to set the zoom on a complete different area that is NOT being viewed and that *shouldn't* prompt a redraw...

SlySven commented:
Not redrawing map after using single argument `setMapZoom(...)`: Confirmed :lady_beetle: .

Strange thing is that the Lua `updateMap()` ***is*** called (in all cases where the setting is successful yet it is proving ineffectual) furthermore that should only be called in the single argument case or when the area being viewed IS the area given in the two argument case...

I need to dig deeper.

Kebap commented:
Please convert to draft until new tests are possible

SlySven committed:
Fix: get setMapZoom(...) working for single argument case

Also revise some error/warning texts.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
> Please convert to draft until new tests are possible

Not really necessary - I think it is fixed-up now. Both the messages and the single argument *setting* case.

SlySven committed:
Merge branch 'development' into Improve_rememberMapperZoomAmountBetweenSessions_simplified

SlySven commented:
:books: Documentation pending in Area 51:
* **[`getMapZoom()`](https://wiki.mudlet.org/index.php?title=Area_51&section=6#getMapZoom.2C_PR_.236615)**
* **[`setMapZoom()`](https://wiki.mudlet.org/index.php?title=Area_51&section=6#setMapZoom.2C_PR_.236615)**
* **~~mapAreaViewedChangeEvent~~ changed to: [`sysMapAreaChanged`](https://wiki.mudlet.org/w/Area_51#mapAreaViewedChangeEvent.2C_PR_.236615)**

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -198,28 +198,49 @@ void T2DMap::slot_switchArea(const QString& newAreaName)
+                // this case we should just use the zoom value that we were
+                // using in the area we are switching from:
+                xyzoom = mpMap->mpRoomDB->get2DMapZoom(areaID);
+                areaViewedChangedEvent.mArgumentList.append(qsl("mapAreaViewedChangeEvent"));
```
I know we went off track in keeping consistency for Mudlet's event names - suggest to bring it back:

```suggestion
                areaViewedChangedEvent.mArgumentList.append(qsl("sysMapAreaChanged"));
```

SlySven replied:
:thinking: 
Of the **64** events Mudlet currently raises, **17** end with `"Event"` (and **7** with `"Error"`), **2** start with `"map"`, **59** start with `"sys"`... so I guess this makes sense and I will do this.

OTOH though, ***I*** do like the "Event" suffix as it gives a clear indication what the string represents when a handler function gets it as the first argument!

SlySven replied:
Done in b307d19ed1549c9c5290e05b6210bf4f0fd54088 (I forgot it when doing the others...)

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -1270,14 +1296,14 @@ void T2DMap::paintEvent(QPaintEvent* e)
          */
         int maxUsedRoomId = 0;
-        QSetIterator<int> itRoomId(playerArea->getAreaRooms());
+        QSetIterator<int> itRoomId(pArea->getAreaRooms());
```
I prefer a more descriptive variable name here - it's not super long either.

SlySven replied:
Yeah, but in reducing the number of `TArea*` instances in the event it turns out that this pointer does not point necessarily to the area the player is ***in***, but the one being painted - so the original name A) did not indicate it was a pointer or B) reflect what it contains.

vadi2 replied:
OK

vadi2 replied:
Someone will make the same mistake later on then though, as `pArea` semantically means nothing. `pDrawnArea`?

SlySven replied:
To me, `pArea` means "a pointer to a TArea instance"... :man_shrugging:  - but I guess I can make that change.

Done in 7e8eb155c37ee41962b501999ded85a615c62c98

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -4482,14 +4514,51 @@ void T2DMap::wheelEvent(QWheelEvent* e)
 }
 
-void T2DMap::setMapZoom(qreal zoom)
+QString T2DMap::setMapZoom(const qreal zoom, const int areaId)
```
Just a suggestion, have you considered using std::pair<bool, QString> as a pattern to return a success boolean and an optional error message? It could be helpful in ensuring consistency and clarity throughout the codebase.

SlySven replied:
:man_shrugging: Yeah I did, but the additional complexity in processing the returned **error-only** messages didn't warrant it - if there is no message then things worked; should the situation change such that something worked partly but not completely or that success also warranted some informational text then that would be a suitable methodology.

SlySven replied:
I mean I could also have used a `std::optional<const char*>` or `std::optional<QString>`...

vadi2 replied:
Why not stay with the consistent approach?

SlySven replied:
Okay, will do though it seems unnecessary...!

Done in 7e8eb155c37ee41962b501999ded85a615c62c98

vadi2 commented on src/T2DMap.h:
```diff
@@ -91,9 +91,13 @@ class T2DMap : public QWidget
 #endif
 
 
+    inline static const qreal csmDefaultXYZoom = 20.0;
+    inline static const qreal csmMinXYZoom = 3.0;
```
```suggestion
    // default 2D zoom level
    inline static const qreal csmDefaultXYZoom = 20.0;
    // minimum 2D zoom level
    inline static const qreal csmMinXYZoom = 3.0;
```

SlySven replied:
Well it is **in** the `T2DMap` class (so other classes will access them as `T2DMap::csmDefaultXYZoom` and `T2DMap::csmMinXYZoom`) so isn't the naming clear enough already? :confused:

vadi2 replied:
It can help in IDEs which show documentation as comments for variables

SlySven replied:
**TIL** that some IDES do ***that***! Actually it sounds a bit like how the Qt translation tools look for specially formatted comments before translations are used. Okay, will do.

vadi2 replied:
CLion does it - https://www.jetbrains.com/help/clion/viewing-inline-documentation.html (you should try it!)

SlySven replied:
Done in 7e8eb155c37ee41962b501999ded85a615c62c98

SlySven committed:
Revise: some changes suggested in peer review

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: another change suggested in peer review

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
:books: Documentation reordered/improved.

SlySven committed:
Merge branch 'development' into Improve_rememberMapperZoomAmountBetweenSessions_simplified
